// areaazul/models/collections.js
module.exports.configuracao = require('./collections/configuracao');
module.exports.consumo = require('./collections/consumo');
module.exports.contrato = require('./collections/conta');
module.exports.credenciado = require('./collections/contrato');
module.exports.fiscal = require('./collections/credenciado');
module.exports.funcionario = require('./collections/fiscal');
module.exports.movimentacaoconta = require('./collections/funcionario');
module.exports.pessoa = require('./collections/movimentacaoconta');
module.exports.pessoafisica = require('./collections/pessoa');
module.exports.pessoajuridica = require('./collections/pessoajuridica');
module.exports.revendedor = require('./collections/revendedor');
module.exports.usuario = require('./collections/usuario');
module.exports.veiculo = require('./collections/veiculo');
module.exports.estado = require('./collections/estado');
module.exports.ativacao = require('./collections/ativacao');
module.exports.fiscalizacao = require('./collections/fiscalizacao');
